runtime/src/main/java/org/corfudb/runtime/LogReplicationUtils.java

```diff
@@ -1,15 +1,160 @@
+                        // cache, a trimmed exception is thrown and requires a retry at a later timestamp.
+                        incrementCount(mvoTrimCounter);
+                        log.warn("Snapshot no longer available in the cache.  Retrying.", tae);
+                    } else if (tae.getAbortCause() == AbortCause.CONFLICT) {
+                        // Concurrent updates to the client's tables
+                        incrementCount(conflictCounter);
+                        log.warn("Concurrent updates to client tables.  Retrying.", tae);
+                    } else {
+                        log.error("Unexpected type of Transaction Aborted Exception", tae);
+                    }
+                    throw new RetryNeededException();
+                } catch (Exception e) {
+                    log.error("Unexpected exception type hit", e);
+                    throw new RetryNeededException();
+                }
+            }).run();
+        } catch (InterruptedException e) {
+            throw new StreamingException(e);
+        } finally {
+            MicroMeterUtils.time(subscribeTimer, "logreplication.subscribe.duration");
+        }
+    }
+
+
+    private static Table<ReplicationStatusKey, ReplicationStatusVal, Message> openReplicationStatusTable(CorfuStore corfuStore) {
+        try {
+            return corfuStore.openTable(CORFU_SYSTEM_NAMESPACE, REPLICATION_STATUS_TABLE, ReplicationStatusKey.class,
+                    ReplicationStatusVal.class, null, TableOptions.fromProtoSchema(ReplicationStatusVal.class));
+        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+            log.error("Failed to open the Replication Status table", e);
+            throw new StreamingException(e);
+        }
+    }
+
+    private static void incrementCount(Optional<Counter> counter) {
+        counter.ifPresent(Counter::increment);
+    }
+
+    /**
+     * If full sync on client tables was not performed during subscription, attempt to perform it now and complete
+     * the subscription.
+     * @param corfuStore
+     * @param clientListener
+     * @param namespace
+     */
+    public static void attemptClientFullSync(CorfuStore corfuStore, LogReplicationListener clientListener,
+                                        String namespace) {
+        long subscriptionTimestamp = getSubscriptionTimestamp(corfuStore, namespace, clientListener);
+        log.info("Client full sync completed at timestamp {}", subscriptionTimestamp);
+        clientListener.getClientFullSyncTimestamp().set(subscriptionTimestamp);
+        updateListenerFlagsForSnapshotSync(clientListener, false);
+    }
+
+    private static void updateListenerFlagsForSnapshotSync(LogReplicationListener clientListener,
+                                                           boolean snapshotSyncInProgress) {
+        clientListener.getClientFullSyncPending().set(snapshotSyncInProgress);
```

Comment:
Do you think a log statement here would help in debugging update listener flags?